src/fs/transaction.ts

```diff
@@ -0,0 +1,590 @@
+import { CID } from "../common/cid.js"
+import { MountedPrivateNodes, PrivateNodeQueryResult } from "./types/internal.js"
+import { Partition, Partitioned, PartitionedNonEmpty, Private, Public } from "../path/index.js"
+import { RootTree } from "./rootTree.js"
+import { Rng } from "./rng.js"
+import { Ucan } from "../ucan/types.js"
+import { agent } from "../did/local.js"
+import { dataFromBytes, dataToBytes } from "./data.js"
+import { findPrivateNode, partition as determinePartition } from "./mounts.js"
+import { addOrIncreaseNameNumber, pathSegmentsWithoutPartition, privateReferenceFromWnfsRef, searchLatest, wnfsRefFromPrivateReference } from "./common.js"
+import { throwNoAccess } from "./errors.js"
+import { hasProp } from "../common/type-checks.js"
+
+
+export class TransactionContext {
+
+  private changedPaths: Set<Path.Distinctive<Partitioned<Partition>>>
+
+  constructor(
+    private blockStore: BlockStore,
+    private dependencies: Dependencies,
+    private privateNodes: MountedPrivateNodes,
+    private rng: Rng,
+    private rootTree: RootTree
+  ) {
+    this.changedPaths = new Set()
+  }
+
+
+  static async commit(context: TransactionContext): Promise<{
+    changedPaths: Array<Path.Distinctive<Partitioned<Partition>>>,
+    privateNodes: MountedPrivateNodes,
+    proofs: Ucan[],
+    rootTree: RootTree,
+  }> {
+    const changedPaths = Array.from(context.changedPaths)
+    const audience = await agent(context.dependencies.crypto)
+
+    const newForest = await changedPaths.reduce(
+      async (oldForestPromise, changedPath): Promise<PrivateForest> => {
+        const oldForest = await oldForestPromise
+
+        if (!Path.isPartition("private", changedPath)) {
+          return oldForest
+        }
+
+        const nodePath = Path.removePartition(changedPath)
+        const posixPath = Path.toPosix(nodePath, { absolute: true })
+        const maybeNode = context.privateNodes[ posixPath ]
+
+        if (maybeNode) {
+          const [ _, newForest ] = await maybeNode.node.store(oldForest, context.blockStore, context.rng)
+          return newForest
+        } else {
+          return oldForest
+        }
+      },
+      Promise.resolve(
+        context.rootTree.privateForest
+      )
+    )
+
+    // Replace forest
+    const rootTree = { ...context.rootTree, privateForest: newForest }
+
+    // Proofs
+    const proofs = await changedPaths.reduce(
+      async (accPromise: Promise<Ucan[]>, changedPath: Path.Distinctive<Partitioned<Partition>>): Promise<Ucan[]> => {
+        const acc = await accPromise
+
+        const proof = await context.dependencies.reference.repositories.ucans.lookupFileSystemUcan(
+          audience,
+          changedPath
+        )
+
+        // Throw error if no write access to this path
+        if (!proof) throwNoAccess(
+          changedPath,
+          "write"
+        )
+
+        return acc
+      },
+      Promise.resolve([])
+    )
```

Comment:
It does check it yeah, although not in a "standard conditional checking" way. The `reduce` call makes it so that `newForest` is actually the old private forest if there are no changes.

Reply:
Yeye absoluetly, what I'm trying to say is take the code I commented on and put it at the start of the function. Should do the same thing, but be faster in the case that you don't actually have write access.